### DISCUSSION TITLE
Give us a simplified API to work with

### DISCUSSION THREAD

**Discussion by vesper8** [opcodesio/log-viewer#94] [Ideas]
I'm migrating from arcanedev/log-viewer because I feel like your package offers extra functionality and is more modern overall and of course very actively developed which is a big plus.

Previously, I had setup some custom commands for checking the status of logs, as well as clearing older logs.

The "log checking" command worked like this:

Check today's log file and check if the number of errors are above a certain threshold, if so, then trigger notifications/emails and include the count and summary of the errors.

I've been looking into doing the same thing using your package and I've been able to get it mostly working, but it's not exactly clean.

Here's some of what I'm currently working on:

```

        $logList = new \Opcodes\LogViewer\Http\Livewire\LogList;

        // get list of files
        $files = \Opcodes\LogViewer\Facades\LogViewer::getFiles();

        // todo, figure out which of the files is the latest log

        // get the latest log file (hardcoded for now)
        $file = \Opcodes\LogViewer\Facades\LogViewer::getFile('laravel-2022-09-08.log');

        // set the levels you're interested in tracking
        $selectedLevels = [
            // 'debug',
            // 'info',
            // 'notice',
            'warning',
            'error',
            'critical',
            'alert',
            'emergency',
            'processing',
            'processed',
            'failed',
            // '',
          ];    

        $logQuery = $file?->logs()->only($selectedLevels);

        $levels = $logQuery?->getLevelCounts();

        $logs = $logQuery?->paginate($logList->perPage);

        // WIP, get a summary of levels included in the log
        foreach ($levels as $key => $value) {
            if ($value->count > 0) {
                $this->info(sprintf(
                    '%s: %d',
                    $key,
                    $value->count
                ));
            }
        }

        // WIP, iterate over the actual log entries and format into a short and neat summary
        foreach ($logs as $index => $log) {
            $this->info(sprintf(
                '[%s] %s: %s',
                $log->time->toDateTimeString(),
                $log->level->getName(),
                $log->text
            ));
        }

        // decide whether the error threshold has been surpassed
        // trigger notifications/emails with summary
```

It would be nice if there was an API that would help with some of the above tasks. I'd like to be able to do something like this:

```

$latestLogFile = (new LogViewer)->latest();

$errorLogs = $latestLogFile->level('error')->logs();

$count = $errorLogs->count();
$logEntries = $errorLogs->entries();

```

Anyway something like this.. I think you get the idea of what I'm trying to achieve. The above code is working but it could break at any time if you make changes and it's a bit hacky.

Of course it would be even more amazing if you had a built-in "log alert" system that could be customized.

Similarly a built-in logrotate that deletes logs older than X days would also be great. It could be as simple as calling a scheduled "log cleanup" command that could be customized with "X days" as a parameter with a sensible default.

In the meantime though an API for clearing/deleting logs would be appreciated.

**Comment by arukompas** [comment]
That's a really good point @vesper8 ! 👏 

I did not think about the API provided while building this tool, but I'll definitely keep it in mind for future updates.

I've made a few small improvements in `v1.4.0` and here's how you can achieve what you need right now:

```php
use Opcodes\LogViewer\Facades\LogViewer;
use Opcodes\LogViewer\Level;

$latestLogFile = LogViewer::getFiles()->latest();

$errorLogs = $latestLogFile->logs()->only(Level::Error);

$count = $errorLogs->count();
$logEntries = $errorLogs->get();   // this will load all related logs into memory. BE CAREFUL! Alternative below

// A better, less memory-intensive approach is to use the generator provided and read one log at a time
while ($log = $errorLogs->next()) {
    // do something with the log returned
}

// Alternatively, you can fetch X number of logs at a time without resetting the cursor
while ($logs = $errorLogs->get(100)) {
    // $logs is now a collection of 100 logs.
    // The next time this gets called, it will be a collection of the NEXT 100 logs.
}
```

**Reply by vesper8** [reply]
That's fantastic!! Thanks for the quick turnaround!! So nice!
